sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -441,8 +406,7 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
 
   override def children: Seq[Expression] = key +: branches
 
-  override lazy val resolved: Boolean =
-    childrenResolved && valueTypesEqual
+  override protected def checkTypesInternal(): TypeCheckResult = TypeCheckResult.success
```

Comment:
do we add the cast elsewhere? if we do, then i think it is ok to not check here.